Add minimal skeleton for _C type stubs, delete torch.autograd stub

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38176 Move torch/autograd/grad_mode.pyi stubs inline
* #38173 Device and torch._C function cleanup
* #38157 Delete torch/__init__.pyi, deferring to direct extension stubs
* #38136 Bind VariableFunctions as a module, not a class with static methods.
* **#38080 Add minimal skeleton for _C type stubs, delete torch.autograd stub**

Originally, my plan was to just delete the torch.autograd stub, but
this triggered a bunch of downstream errors relating to non-existent
to _C modules, and so instead of ignoring those files, I decided to
add a minimal _C type stubs, where it was easy (cases which were
codegened I ignored).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21487841](https://our.internmc.facebook.com/intern/diff/D21487841)

ezyang committed:
Add minimal skeleton for _C type stubs, delete torch.autograd stub

Originally, my plan was to just delete the torch.autograd stub, but
this triggered a bunch of downstream errors relating to non-existent
to _C modules, and so instead of ignoring those files, I decided to
add a minimal _C type stubs, where it was easy (cases which were
codegened I ignored).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Add minimal skeleton for _C type stubs, delete torch.autograd stub"


Originally, my plan was to just delete the torch.autograd stub, but
this triggered a bunch of downstream errors relating to non-existent
to _C modules, and so instead of ignoring those files, I decided to
add a minimal _C type stubs, where it was easy (cases which were
codegened I ignored).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

rgommers commented on torch/autograd/__init__.py:
```diff
@@ -50,7 +51,15 @@ def _make_grads(outputs, grads):
 
 
-def backward(tensors, grad_tensors=None, retain_graph=None, create_graph=False, grad_variables=None):
+_TensorOrTensors = Union[torch.Tensor, Sequence[torch.Tensor]]
```
This one may be worth moving to `torch.types` when we add that (soon).

ezyang replied:
I'll go ahead and add this right now

rgommers commented on torch/autograd/__init__.py:
```diff
@@ -50,7 +51,15 @@ def _make_grads(outputs, grads):
+    tensors: _TensorOrTensors,
+    grad_tensors: Optional[_TensorOrTensors] = None,
+    retain_graph: Optional[bool] = None,
+    create_graph: bool = None
```
You lost `grad_variables` here, it's not documented but still used in the function body.

rgommers commented on torch/autograd/function.py:
```diff
@@ -347,6 +348,8 @@ def unflatten_helper(input, proto):
 
 class NestedIOFunction(Function):
+    # The 'type: ignore' statements are needed here because these functions are declared as '@staticmethod' in the
+    # superclass (Function) but are instance methods here, which mypy reports as incomptabile.
```
Hmm, mypy 0.770 seems happy without the ignores. What's the error you're seeing?

rgommers replied:
Also, typo in `incompatible`

ezyang replied:
I cargo culted this from the old pyi file. I'll go ahead and remove it.

rgommers commented on torch/autograd/function.py:
```diff
@@ -364,21 +367,21 @@ def _do_backward(self, gradients, retain_variables):
     __call__ = _do_forward
 
-    def forward(self, *args):
+    def forward(self, *args: Any) -> None:  # type: ignore
```
returns a tuple of `<stuff>`, so `-> Any`?

rgommers commented on torch/autograd/gradcheck.py:
```diff
@@ -195,7 +196,22 @@ def _differentiable_outputs(x):
+# the '...' first argument of Callable can be replaced with VarArg(Tensor).
+# For now, we permit any input.
+def gradcheck(
+    func: Callable[..., Union[torch.Tensor, Tuple[torch.Tensor, ...]]],
```
Takes a list of tuples as well (see `_as_tuple` call in function body); did you type this as `Tuple` instead of `Sequence` on purpose? If not, could use the same `_TensorOrTensors` you defined in `__init__.py`.

rgommers replied:
or maybe that's related to the _vararg of tensors, which isn't expressable_ comment?

ezyang replied:
I think the Vararg of tensors is referring to the `...` in Callable

rgommers commented on mypy.ini:
```diff
@@ -193,6 +193,9 @@ ignore_errors = True
 [mypy-torch.utils.bundled_inputs]
 ignore_errors = True
 
```
Can you remove
```
[mypy-torch._C]
ignore_missing_imports = True
```
from this file?

rgommers commented on torch/autograd/gradcheck.py:
```diff
@@ -337,9 +353,16 @@ def fn(input):
-                  nondet_tol=0.0):
+def gradgradcheck(
+    func: Callable[..., Union[torch.Tensor, Tuple[torch.Tensor, ...]]],
+    inputs: Union[torch.Tensor, Tuple[torch.Tensor, ...]],
```
You lost `grad_outputs` here.

rgommers reviewed (commented):
Overall looks good to me, all comments are minor issues.

ezyang committed:
Update on "Add minimal skeleton for _C type stubs, delete torch.autograd stub"


Originally, my plan was to just delete the torch.autograd stub, but
this triggered a bunch of downstream errors relating to non-existent
to _C modules, and so instead of ignoring those files, I decided to
add a minimal _C type stubs, where it was easy (cases which were
codegened I ignored).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

rgommers reviewed (approved):
LGTM now. +1 for `torch.types` :)

ezyang committed:
Update on "Add minimal skeleton for _C type stubs, delete torch.autograd stub"


Originally, my plan was to just delete the torch.autograd stub, but
this triggered a bunch of downstream errors relating to non-existent
to _C modules, and so instead of ignoring those files, I decided to
add a minimal _C type stubs, where it was easy (cases which were
codegened I ignored).

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]
